service/matching/version_sets.go

```diff
@@ -0,0 +1,209 @@
+}
+
+//nolint:revive // cyclomatic complexity
+func updateImpl(existingData *persistence.VersioningData, req *workflowservice.UpdateWorkerBuildIdCompatabilityRequest) error {
+	// First find if the targeted version is already in the sets
+	targetedVersion := extractTargetedVersion(req)
+	findRes := findVersion(existingData, targetedVersion)
+	targetSetIx, versionInSetIx := findRes.setIx, findRes.indexInSet
+
+	if req.GetAddNewBuildIdInNewDefaultSet() != "" {
+		// If it's not already in the sets, add it as the new default set
+		if targetSetIx != -1 {
+			return serviceerror.NewInvalidArgument(fmt.Sprintf("version %s already exists", targetedVersion))
+		}
+
+		existingData.VersionSets = append(existingData.GetVersionSets(), &taskqueuepb.CompatibleVersionSet{
+			VersionSetId: uuid.New(),
+			BuildIds:     []string{targetedVersion},
+		})
+	} else if addNew := req.GetAddNewCompatibleBuildId(); addNew != nil {
+		compatVer := addNew.GetExistingCompatibleBuildId()
+		compatSetIx := findVersion(existingData, compatVer).setIx
+		if compatSetIx == -1 {
+			return serviceerror.NewNotFound(
+				fmt.Sprintf("targeted compatible_version %v not found", compatVer))
+		}
+		if targetSetIx != -1 {
+			// If the version does exist, this operation can't do anything meaningful, but we can fail if the user
+			// says the version is now compatible with some different set.
+			return serviceerror.NewInvalidArgument(fmt.Sprintf("version %s already exists", targetedVersion))
+		}
+
+		// If the version doesn't exist, add it to the compatible set
+		existingData.VersionSets[compatSetIx].BuildIds =
+			append(existingData.VersionSets[compatSetIx].BuildIds, targetedVersion)
+		if addNew.GetMakeSetDefault() {
+			makeDefaultSet(existingData, compatSetIx)
+		}
+	} else if req.GetPromoteSetByBuildId() != "" {
+		if targetSetIx == -1 {
+			return serviceerror.NewNotFound(fmt.Sprintf("targeted version %v not found", targetedVersion))
+		}
+		makeDefaultSet(existingData, targetSetIx)
+	} else if req.GetPromoteBuildIdWithinSet() != "" {
+		if targetSetIx == -1 {
+			return serviceerror.NewNotFound(fmt.Sprintf("targeted version %v not found", targetedVersion))
+		}
+		makeVersionInSetDefault(existingData, targetSetIx, versionInSetIx)
+	}
+
+	return nil
+}
+
+func extractTargetedVersion(req *workflowservice.UpdateWorkerBuildIdCompatabilityRequest) string {
+	if req.GetAddNewCompatibleBuildId() != nil {
+		return req.GetAddNewCompatibleBuildId().GetNewBuildId()
+	} else if req.GetPromoteSetByBuildId() != "" {
+		return req.GetPromoteSetByBuildId()
+	} else if req.GetPromoteBuildIdWithinSet() != "" {
+		return req.GetPromoteBuildIdWithinSet()
+	}
+	return req.GetAddNewBuildIdInNewDefaultSet()
+}
+
+type findVersionRes struct {
+	setIx      int
+	indexInSet int
+}
+
+// Finds the version in the version sets, returning (set index, index within that set)
+// Returns -1, -1 if not found.
+func findVersion(data *persistence.VersioningData, buildID string) findVersionRes {
```

Comment:
use multiple return values instead of struct?
```suggestion
func findVersion(data *persistence.VersioningData, buildID string) (setIndex int, indexInSet int) {
```

Reply:
The linter bitched at me for that lol